File: Ch-1-Arrays/q8.java
/*
Q - 8: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and
column are set to 0.
*/

import java.util.*;

public class q8{
	private void convertMatrix(int[][] a, int r, int c){
		System.out.println();
		int[] rows = new int[r], cols = new int[c];
		for(int i = 0;i < r;i++){
			for(int j = 0;j < c;j++){
				if(a[i][j] == 0){
					rows[i] = 1;
					cols[j] = 1;
				}
			}
		}

		for(int i = 0;i < r;i++){
			for(int j = 0;j < c;j++){
				if(rows[i] == 1 || cols[j] == 1){
					a[i][j] = 0;
				}
			}
		}
	}

	private void displayMatrix(int[][] a, int r, int c){
		for(int i = 0;i < r;i++){
			for(int j = 0;j < c;j++){
				System.out.print(a[i][j]+" ");
			}
			System.out.println();
		}
	}

	public static void main(String args[]){
		int[][] a = {{2,3,5,0},{2,0,3,4},{1,2,3,4},{7,2,4,0},{1,2,3,4}};
		int r = 5, c = 4;
		q8 obj = new q8();
		obj.displayMatrix(a,r,c);
		obj.convertMatrix(a,r,c);
		obj.displayMatrix(a,r,c);
	}
}

File: Ch-1-Arrays/q8.py
'''
Q - 8: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and
column are set to 0.
'''

def convertMatrix(a,r,c):
	rows = []
	cols = []
	for i in range(len(a)):
		for j in range(len(a[0])):
			if a[i][j] == 0:
				rows.append(i)
				cols.append(j)
	for i in range(len(a)):
		for j in range(len(a[0])):
			if i in rows or j in cols:
				a[i][j] = 0

a = [[1,2,3,4],[1,2,0,4],[9,0,3,5],[7,4,5,1],[0,2,4,1]]

print a

convertMatrix(a,5,4)

#the value of a in changed as it is paased by reference in python as well
print a
